src/titiler/extensions/titiler/extensions/render.py

```diff
@@ -0,0 +1,182 @@
+"""STAC Render Extension.
+
+Implements support for reading and applying Item level render extension.
+See: https://github.com/stac-extensions/render
+"""
+
+from typing import Dict, List, Optional
+from urllib.parse import urlencode
+
+from attrs import define
+from fastapi import Depends, HTTPException, Path, Request
+from fastapi.dependencies.utils import get_dependant, request_params_to_args
+from pydantic import BaseModel
+from typing_extensions import Annotated
+
+from titiler.core.factory import FactoryExtension, MultiBaseTilerFactory
+from titiler.core.models.OGC import Link
+
+
+class RenderItem(BaseModel, extra="allow"):
+    """Render item for stac render extension."""
+
+    assets: List[str]
+    title: Optional[str] = None
+    rescale: Optional[List[Annotated[List[float], 2]]] = None
+    nodata: Optional[float] = None
+    colormap_name: Optional[str] = None
+    colormap: Optional[Dict] = None
+    color_formula: Optional[str] = None
+    resampling: Optional[str] = None
+    expression: Optional[str] = None
+    minmax_zoom: Optional[Annotated[List[int], 2]] = None
+
+
+class RenderItemWithLinks(BaseModel):
+    """Same as RenderItem with url and params."""
+
+    params: RenderItem
+    links: List[Link]
+
+
+class RenderItemList(BaseModel):
+    """List of Render Items with links."""
+
+    renders: Dict[str, RenderItemWithLinks]
+    links: List[Link]
+
+
+@define
+class stacRenderExtension(FactoryExtension):
+    """Add /renders endpoint to a STAC TilerFactory."""
+
+    def register(self, factory: MultiBaseTilerFactory):
+        """Register endpoint to the tiler factory."""
+
+        def _prepare_query_string(render: Dict, src_path: str) -> str:
+            """Prepare render related query params.
+
+            Validates and filters query params that titiler can understand.
+            If titiler does not support parameter, it will be ignored.
+            """
+            # List of dependencies a `/tile` URL should validate
+            # Note: Those dependencies should only require Query() inputs
+            tile_dependencies = [
+                factory.layer_dependency,
+                factory.dataset_dependency,
+                # Image rendering Dependencies
+                factory.rescale_dependency,
+                factory.color_formula_dependency,
+                factory.colormap_dependency,
+                factory.render_dependency,
+            ]
+
+            query = {"url": src_path}
+            for dependency in tile_dependencies:
+                dep = get_dependant(path="", call=dependency)
+                if dep.query_params:
+                    # call the dependency with the query-parameters values
+                    query_values, _errors = request_params_to_args(
+                        dep.query_params, render
+                    )
+                    _ = dependency(**query_values)
+                    query.update(query_values)
```

Comment:
what do you think about moving this outside the extensions package to titiler.core.utils? 

so we can replace reuse this outside (in titiler-pgstac and titiler-stacapi) 

https://github.com/developmentseed/titiler-stacapi/blob/270c08d7ebbb26d48bddd9c573957cbf7641da7b/titiler/stacapi/factory.py#L74-L90

Reply:
ouch, I started doing it and found out that all three use cases are slightly different.

The difference between usage here and in titiler-fastapi is evident on `BidxParams`:
```
get_dependency_params(
    dependency=BidxParams, query_params={"bidx": [1, 2, 3]}
)
```
 titiler-fastapi wants to receive `{"index": [1,2,3]}`, I need `{"bidx": [1,2,3]}`, and the key here is in `alias` - [code](https://github.com/alekzvik/titiler/blob/feat-stac-render-extension/src/titiler/core/titiler/core/dependencies.py#L99)
 
 Each project is doing a slightly different task:
 titiler-pgstac - make sure that passed params pass validation for query params
 titiler-fastapi - validate and conform passed params based on query params
 render extension - filter passed params to the ones that can be in query params and make sure they validate
 
I can write a single function to cover two use cases, but not three. I will think about it a bit more and see if something comes up

Reply:
I agree, will move to utils.

One more note, initial implementation in render extension, does not only validate but also picks (or filters) only params that would be valid in the tiling endpoint. For example, if we have in render item of the stac `{"bidx": [1, 2, 3], "other": "value"}`, now I would pass on to the tiling endpoint only `{"bidx": [1, 2, 3]}`.

I prefer this approach because we are returning a URL we can guarantee will work. Otherwise, we can pass on a param that will fail on the validation, e.g. `tilescale=5`.

I am bringing this up to make sure we are making the decision knowing the details.

Reply:
I moved it out to utils, but I also understood that it works a bit differently than I thought before. I will add some more comments near the relevant code to explain more.